File: project-cam-desktop/src/blm.ts
// The launcher console's intent vocabulary. Mirrors ConsoleCommand in
// src-tauri/src/blm.rs (contract-tested in tests/test_desktop_launcher_console.py).
//
// The UI never composes serial text and never composes the bridge's protocol
// text: it names one of these and the backend renders the line. Ranges here are
// for the WIDGETS — the backend refuses out-of-range values and the Python
// bridge clamps as a last line of defence, so a slider bug cannot become a
// different shot than the operator asked for.

export const ANGLE_LIMIT_DEG = 30;
/** The DEFAULT pitch envelope a session starts with — conservative, since no
 *  downward travel cannot jam the barrel against the ball feeder.
 *
 *  NOT a fixed limit. The collision is at a fixed physical position while the
 *  firmware's angle is measured from a zero adopted at boot or by SET ZERO, so a
 *  constant in that frame points somewhere else after every re-zero — and it
 *  deletes legitimate downward travel, since how much room is left below zero
 *  depends on where zero was set. The live envelope comes from the console
 *  status; these are only the fallback before any status arrives. */
export const PITCH_DEFAULT_MIN_DEG = 0;
export const PITCH_DEFAULT_MAX_DEG = 30;
export const YAW_LIMIT_DEG = 30;
export const RPM_MAX = 1200;
/** The firmware refuses `shoot` below this, and the bridge refuses to arm. */
export const RPM_MIN_FIRE = 400;
/** A fire request must carry exactly this word, so a stray invoke is not a shot. */
export const FIRE_CONFIRMATION = "FIRE";

export type FitKind = "linear" | "quadratic" | "interp";

export type ConsoleCommand =
  | { command: "aim"; pitch_deg: number; yaw_deg: number; wheel_rpm: number }
  | { command: "wheels"; wheel_rpm: number }
  | { command: "reload" }
  | { command: "arm" }
  | { command: "disarm" }
  | { command: "fire"; confirm: string }
  | { command: "stop" }
  | { command: "clear" }
  | { command: "center" }
  /** Adopt the barrel's current physical position as zero. Moves nothing. */
  | { command: "set_zero" }
  /** Declare the pitch travel measured from the current zero. */
  | { command: "limits"; pitch_min_deg: number; pitch_max_deg: number }
  | { command: "info" }
  /** No rpm by design: the wheels are stopped before a distance can be measured,
   *  so the bridge takes the RPM from the shot that was actually fired. */
  | { command: "measure"; landing_distance_m: number }
  | { command: "undo" }
  | { command: "fit"; height_m: number; kind: FitKind };

/** A serial device the backend found, with the evidence for what it is.
 *
 *  Mirrors `SerialDevice` in src-tauri/src/launch_profiles.rs. Detection is
 *  PASSIVE — USB identity read from sysfs, no port opened — hence
 *  `likely_launcher` rather than `is_launcher`: only opening the console and
 *  polling the firmware proves what is on the other end, and the UI must not
 *  imply otherwise.
 */
export type SerialDevice = {
  /** What to launch with: a stable /dev/serial/by-id link when one exists. */
  path: string;
  /** The kernel node it currently resolves to — this is what moves on replug. */
  node: string;
  label: string;
  usb_id: string;
  likely_launcher: boolean;
  reason: string;
};

/** Pick the device to preselect, without ever overriding a deliberate choice.
 *
 *  `current` is kept whenever it is still present, so a manual override survives
 *  the periodic refresh. Otherwise the likely launcher wins — the list is already
 *  sorted with those first, so this is the first entry in practice.
 */
export function autoSelectPort(devices: SerialDevice[], current: string): string {
  if (current && devices.some((device) => device.path === current)) return current;
  const preferred = devices.find((device) => device.likely_launcher) ?? devices[0];
  return preferred?.path ?? "";
}

/** Status the bridge publishes on stdout, one line per change, prefixed so the
 *  mission log can route it out of the human-readable stream. */
export const STATUS_PREFIX = "@BLM ";

export type ConsoleStatus = {
  schema: string;
  port: string;
  connected: boolean;
  allow_fire: boolean;
  estop_latched: boolean;
  armed: boolean;
  arm_remaining_s: number;
  arm_timeout_s: number;
  pitch_deg: number | null;
  yaw_deg: number | null;
  /** An aim the operator actually established. `pitch_deg` alone is not that: an
   *  RPM-only change also has to state the angles, because the firmware takes one
   *  combined `set v h wl wr`. */
  aim_established: boolean;
  wheel_rpm: number;
  /** MEASURED flywheel RPM, and how old the reading is. The age is not optional
   *  decoration: the firmware stops sending while the pusher moves and a dead
   *  reader leaves the last numbers in place forever, so a frozen "0 / 0" would
   *  read as "the wheels are stopped, it is safe to walk out". `null` age means no
   *  reading has ever arrived. */
  rpm_left: number | null;
  rpm_right: number | null;
  telemetry_age_s: number | null;
  telemetry_max_age_s: number;
  /** The bridge's verdicts, NOT recomputed here. These are the same predicates its
   *  arm and fire gates use, and one safety rule must have one implementation. */
  wheels_confirmed: boolean;
  wheels_unconfirmed_reason: string;
  /** The FULL arm predicate: agreement with the command, then enough separate
   *  arrivals, then enough span between them. `wheels_confirmed` alone is
   *  necessary but not sufficient — a panel showing only that would look ready
   *  while ARM refuses. `wheels_in_band_s` is the span the SAMPLES cover, so it
   *  never grows while nothing is arriving. */
  wheels_stable: boolean;
  wheels_unstable_reason: string;
  wheels_sample_count: number;
  wheels_stable_min_samples: number;
  wheels_in_band_s: number;
  wheels_band_rpm: number;
  wheels_stable_required_s: number;
  rpm_spread_max: number;
  /** Whether the MACHINE says the flywheels are stopped: commanded zero, a fresh
   *  reading, and both wheels under the threshold. */
  safe_to_approach: boolean;
  rpm_safe_approach: number;
  /** A reload since the last shot — bookkeeping the console is entitled to. */
  loaded: boolean;
  /** What the last poll's ball switch said. Inferred polarity, so it informs and
   *  never gates; `null` until a poll has been seen. */
  ball_present: boolean | null;
  info_age_s: number | null;
  rpm_min_fire: number;
  angle_limit_deg: number;
  /** The live envelope, in the current zero's frame. */
  pitch_min_deg: number;
  pitch_max_deg: number;
  pitch_default_min_deg: number;
  pitch_default_max_deg: number;
  yaw_limit_deg: number;
  /** Shots the FIRMWARE confirmed, not `shoot` commands written. */
  shots_fired: number;
  shot_ack_timeout_s: number;
  /** A `shoot` that reached the port and has not yet been acknowledged by the
   *  firmware's front limit. While this is non-null the outcome is unknown: no
   *  shot exists, no distance may be recorded, and every command that could
   *  change the physical outcome is refused. `timed_out` means the bridge gave
   *  up and latched STOP — that state is not clearable from the panel. */
  fire_request: {
    request_seq: number;
    rpm: number;
    rpm_left_pre_fire: number;
    rpm_right_pre_fire: number;
    rpm_pre_fire_sample_age_s: number;
    confirmation_age_s: number;
    timed_out: boolean;
  } | null;
  /** The CONFIRMED shot waiting for its landing distance, carrying the RPM
   *  commanded when it was taken. Null when nothing is outstanding, which is
   *  when RECORD SHOT has nothing to attach to.
   *
   *  The RPM pair says `pre_fire` because that is what it is: the firmware gates
   *  telemetry on STATE_IDLE and is in STATE_SHOOTING for the whole
   *  acknowledgement window, so no reading contemporaneous with the shot can
   *  exist. Its age travels with it for the same reason. */
  pending_shot: {
    rpm: number;
    seq: number;
    request_seq: number;
    rpm_left_pre_fire: number;
    rpm_right_pre_fire: number;
    rpm_pre_fire_sample_age_s: number;
  } | null;
  last_refusal: string;
  info_lines: string[];
  measurements: { rpm: number; distance_m: number; shot_seq: number }[];
  model_path: string;
  model_summary: string;
};

/** Parse a `@BLM {...}` line. Returns null for anything else, so an ordinary log
 *  line can never be mistaken for telemetry. */
export function parseStatusLine(line: string): ConsoleStatus | null {
  if (!line.startsWith(STATUS_PREFIX)) return null;
  try {
    const parsed = JSON.parse(line.slice(STATUS_PREFIX.length));
    if (!parsed || typeof parsed !== "object") return null;
    return parsed as ConsoleStatus;
  } catch {
    return null;
  }
}

/** Everything the operator must satisfy before a shot is even offered.
 *
 *  `roomClear` is the one condition no machine can check: the bridge enforces
 *  arm-then-fire, one shot per arm and the RPM gate, but only a person can say
 *  the room is empty. Kept here so the reason a FIRE button is disabled is
 *  always visible rather than inferred.
 */
export function fireBlockers(
  status: ConsoleStatus | null,
  roomClear: boolean
): string[] {
  const blockers: string[] = [];
  if (!status) return ["console is not running"];
  if (!status.connected) blockers.push("serial link is down");
  if (!status.allow_fire) blockers.push("console was opened without fire control");
  if (status.estop_latched) blockers.push("ESTOP latched — release it first");
  if (!roomClear) blockers.push("room-clear not confirmed");
  if (!status.aim_established) blockers.push("no aim sent yet");
  if (status.wheel_rpm < RPM_MIN_FIRE)
    blockers.push(`wheels commanded below the ${RPM_MIN_FIRE} RPM gate`);
  // The commanded RPM says what was asked for; only these two say what happened.
  if (!status.loaded) blockers.push("no reload since the last shot");
  if (!status.wheels_confirmed)
    blockers.push(status.wheels_unconfirmed_reason || "flywheels not confirmed");
  if (!status.armed) blockers.push("not armed");
  return blockers;
}

/** One named step, so the per-shot cycle is on screen instead of on paper.
 *
 *  The cycle is not incidental: firmware `reload` homes both aim axes AND zeroes
 *  the wheel targets, and a shot consumes the arm — so RPM and ARM genuinely have
 *  to be re-established for every single shot. That surprised the operator on the
 *  first pass, which is the tell that the sequence needed to be shown rather than
 *  remembered.
 *
 *  Every branch reads a field the BRIDGE computed. This function orders and names
 *  them; it never decides a safety question of its own.
 */
export type CycleStep = {
  key: string;
  title: string;
  detail: string;
  tone: "idle" | "ask" | "wait" | "ready" | "danger";
};

export function cycleStep(
  status: ConsoleStatus | null,
  roomClear: boolean
): CycleStep {
  if (!status || !status.connected)
    return {
      key: "closed",
      title: "OPEN THE CONSOLE",
      detail: "Pick the launcher port and press OPEN CONSOLE.",
      tone: "idle",
    };
  if (status.estop_latched)
    return {
      key: "latched",
      title: "RELEASE THE LATCH",
      detail: "Every actuating command is refused until the ESTOP latch is released.",
      tone: "danger",
    };

  // A fired shot outranks everything else: the ball is on the floor and the only
  // way to walk out to it is through a confirmed spin-down.
  if (status.pending_shot) {
    const shot = status.pending_shot;
    if (status.wheel_rpm !== 0)
      return {
        key: "spin_down",
        title: "SET RPM TO 0",
        detail: `Shot ${shot.seq} is on the floor. Nobody may walk downrange until the wheels are commanded to zero.`,
        tone: "danger",
      };
    if (!status.safe_to_approach)
      return {
        key: "spinning_down",
        title: "DO NOT APPROACH",
        detail:
          status.telemetry_age_s === null ||
          status.telemetry_age_s > status.telemetry_max_age_s
            ? "The flywheel reading is stale, so the console cannot tell whether they have stopped. Do not use it as permission."
            : `Waiting for both wheels below ${status.rpm_safe_approach} rpm.`,
        tone: "danger",
      };
    return {
      key: "measure",
      title: `MEASURE SHOT ${shot.seq}`,
      detail: `Wheels confirmed stopped. Measure from directly below the barrel to the first floor contact, then RECORD SHOT at ${shot.rpm} rpm.`,
      tone: "ask",
    };
  }

  if (!status.loaded)
    return {
      key: "reload",
      title: "RELOAD",
      detail:
        "Reload also homes the aim and zeroes the wheels, so it comes first — everything after it is set against a known state.",
      tone: "ask",
    };
  if (!status.aim_established)
    return {
      key: "aim",
      title: "SET THE AIM",
      detail: "Commit PITCH and YAW. An RPM-only change does not count as an aim.",
      tone: "ask",
    };
  if (status.wheel_rpm < RPM_MIN_FIRE)
    return {
      key: "spin_up",
      title: "COMMAND THE PASS RPM",
      detail: `The reload zeroed the wheels. Pick the pass RPM — at least ${RPM_MIN_FIRE} to clear the firmware gate.`,
      tone: "ask",
    };
  if (!status.wheels_confirmed)
    return {
      key: "unconfirmed",
      title: "WHEELS NOT CONFIRMED",
      detail: status.wheels_unconfirmed_reason,
      tone: "wait",
    };
  if (status.wheels_in_band_s < status.wheels_stable_required_s)
    return {
      key: "stabilising",
      title: "CONFIRMING THE WHEELS",
      detail: `Held ${status.wheels_in_band_s.toFixed(1)} s of the ${status.wheels_stable_required_s.toFixed(1)} s required inside ±${status.wheels_band_rpm} rpm.`,
      tone: "wait",
    };
  if (!status.allow_fire)
    return {
      key: "aim_only",
      title: "AIM ONLY",
      detail:
        "This console was opened without fire control. Reopen it with ENABLE FIRE CONTROL to arm.",
      tone: "idle",
    };
  if (!roomClear)
    return {
      key: "room",
      title: "CONFIRM THE ROOM",
      detail: "The one condition no machine can check. Nobody downrange.",
      tone: "ask",
    };
  if (!status.armed)
    return {
      key: "arm",
      title: "ARM",
      detail: `Wheels confirmed for ${status.wheels_in_band_s.toFixed(1)} s. The arm lasts ${status.arm_timeout_s} s and one shot consumes it.`,
      tone: "ready",
    };
  return {
    key: "fire",
    title: "HOLD TO FIRE",
    detail: `Armed for ${status.arm_remaining_s.toFixed(0)} s more. Start the slow-motion capture first.`,
    tone: "ready",
  };
}

File: project-cam-desktop/src-tauri/src/blm.rs
//! The launcher console's intent vocabulary.
//!
//! The UI does not write serial, and it does not write the bridge's protocol
//! text either: it names one of these variants and the backend renders the line.
//! That keeps the same property the launch boundary has — there is no variant
//! that can express "send these bytes" — one layer further in, so a compromised
//! or buggy frontend cannot compose a firmware command.
//!
//! Ranges are duplicated from `garage_lab_combined/scripts/blm_bridge.py` on
//! purpose. The bridge is authoritative (it also clamps, because a human piping
//! it by hand is a supported path); this layer REFUSES instead of clamping, so a
//! UI that sends an out-of-range angle produces a visible error rather than a
//! silently different shot.

use serde::Deserialize;

/// Beyond this the ESP32 reboots — see .claude/rules/safety.md "Known Hazards".
pub const ANGLE_LIMIT_DEG: f64 = 30.0;
pub const YAW_LIMIT_DEG: f64 = 30.0;

/// The DEFAULT pitch envelope a session starts with — conservative, because no
/// downward travel cannot jam the barrel against the ball feeder.
///
/// Deliberately not a hard limit here. The collision is at a fixed PHYSICAL
/// position while the firmware's angle is measured from a zero adopted at boot or
/// by `set_zero`, so a constant in this frame means a different physical place
/// after every re-zero. The live envelope is per-session state in
/// `blm_bridge.py`, declared by the operator with `limits`, translated when the
/// zero moves, and enforced there; this layer only guards the firmware bound,
/// which IS frame-independent.
pub const PITCH_DEFAULT_MIN_DEG: f64 = 0.0;
pub const PITCH_DEFAULT_MAX_DEG: f64 = 30.0;

// A declarable limit may only ever sit inside the firmware bound. Widening one
// past it would hand the operator an angle that reboots the ESP32, so this fails
// the build rather than waiting for someone to notice at the bench.
const _: () = assert!(PITCH_DEFAULT_MIN_DEG >= -ANGLE_LIMIT_DEG);
const _: () = assert!(PITCH_DEFAULT_MAX_DEG <= ANGLE_LIMIT_DEG);
const _: () = assert!(YAW_LIMIT_DEG <= ANGLE_LIMIT_DEG);
const _: () = assert!(PITCH_DEFAULT_MIN_DEG < PITCH_DEFAULT_MAX_DEG);
pub const RPM_MAX: f64 = 1200.0;
/// The firmware refuses `shoot` below this; the bridge refuses to arm below it.
///
/// Not referenced by the renderer, and deliberately so: the gate needs the wheel
/// state, which only the bridge holds. It lives here as the Rust side of the
/// three-layer parity check (tests/test_desktop_launcher_console.py) — one value
/// per layer, compared, so the number cannot drift in one of them.
#[allow(dead_code)]
pub const RPM_MIN_FIRE: f64 = 400.0;
/// A fire request must carry this exact word. An empty-payload invoke cannot
/// fire, so a stray call or a mis-wired button is not a shot.
pub const FIRE_CONFIRMATION: &str = "FIRE";

#[derive(Clone, Copy, Debug, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum FitKind {
    Linear,
    Quadratic,
    Interp,
}

impl FitKind {
    fn as_str(self) -> &'static str {
        match self {
            Self::Linear => "linear",
            Self::Quadratic => "quadratic",
            Self::Interp => "interp",
        }
    }
}

#[derive(Clone, Debug, Deserialize)]
#[serde(tag = "command", rename_all = "snake_case", deny_unknown_fields)]
pub enum ConsoleCommand {
    Aim {
        pitch_deg: f64,
        yaw_deg: f64,
        wheel_rpm: f64,
    },
    Wheels {
        wheel_rpm: f64,
    },
    Reload {},
    Arm {},
    Disarm {},
    Fire {
        confirm: String,
    },
    Stop {},
    Clear {},
    Center {},
    /// Adopt the barrel's current physical position as zero. Moves nothing, so it
    /// is not actuation — and it is needed precisely when the reference is wrong.
    SetZero {},
    /// Declare the pitch travel measured from the CURRENT zero. The bridge holds
    /// and enforces it; a re-zero translates the same physical endpoints into
    /// the new coordinate frame.
    Limits {
        pitch_min_deg: f64,
        pitch_max_deg: f64,
    },
    Info {},
    /// Record a method-A landing distance for the v(RPM) fit.
    Measure {
        rpm: f64,
        landing_distance_m: f64,
    },
    Undo {},
    Fit {
        height_m: f64,
        kind: FitKind,
    },
}

fn finite(value: f64, name: &str) -> Result<f64, String> {
    if !value.is_finite() {
        return Err(format!("{name} must be a finite number"));
    }
    Ok(value)
}

/// Guard the FIRMWARE bound only. The mechanical envelope is per-session state
/// that only the bridge holds, and it clamps and reports against it — refusing
/// here on a fixed number would either jam the machine (too wide) or delete
/// legitimate travel (too narrow), depending on where zero happens to sit.
fn pitch(value: f64) -> Result<f64, String> {
    let value = finite(value, "pitch_deg")?;
    if value.abs() > ANGLE_LIMIT_DEG {
        return Err(format!(
            "pitch_deg must be within +/-{ANGLE_LIMIT_DEG} deg (the firmware bound), \
             got {value}"
        ));
    }
    Ok(value)
}

fn yaw(value: f64) -> Result<f64, String> {
    let value = finite(value, "yaw_deg")?;
    if value.abs() > YAW_LIMIT_DEG {
        return Err(format!(
            "yaw_deg must be within +/-{YAW_LIMIT_DEG} deg, got {value}"
        ));
    }
    Ok(value)
}

fn rpm(value: f64) -> Result<f64, String> {
    let value = finite(value, "wheel_rpm")?;
    if !(0.0..=RPM_MAX).contains(&value) {
        return Err(format!(
            "wheel_rpm must be between 0 and {RPM_MAX}, got {value}"
        ));
    }
    Ok(value)
}

impl ConsoleCommand {
    /// Render one line of the bridge protocol, or refuse.
    pub fn render(&self) -> Result<String, String> {
        Ok(match self {
            Self::Aim {
                pitch_deg,
                yaw_deg,
                wheel_rpm,
            } => format!(
                "aim {:.0} {:.0} {:.0}",
                pitch(*pitch_deg)?,
                yaw(*yaw_deg)?,
                rpm(*wheel_rpm)?
            ),
            Self::Wheels { wheel_rpm } => format!("wheels {:.0}", rpm(*wheel_rpm)?),
            Self::Reload {} => "reload".into(),
            Self::Arm {} => "arm".into(),
            Self::Disarm {} => "disarm".into(),
            Self::Fire { confirm } => {
                if confirm != FIRE_CONFIRMATION {
                    return Err(format!("fire requires confirm == {FIRE_CONFIRMATION:?}"));
                }
                "fire".into()
            }
            Self::Stop {} => "stop".into(),
            Self::Clear {} => "clear".into(),
            Self::Center {} => "center".into(),
            // Bridge protocol, not the raw firmware word. Python alone owns the
            // `set_zero` -> `setzero` translation.
            Self::SetZero {} => "set_zero".into(),
            Self::Limits {
                pitch_min_deg,
                pitch_max_deg,
            } => {
                let low = pitch(*pitch_min_deg)?;
                let high = pitch(*pitch_max_deg)?;
                if low >= high {
                    return Err(format!(
                        "pitch_min_deg must be below pitch_max_deg, got {low} >= {high}"
                    ));
                }
                if low > 0.0 || high < 0.0 {
                    return Err(format!(
                        "pitch travel must contain the current zero, got [{low}, {high}]"
                    ));
                }
                format!("limits {low:.0} {high:.0}")
            }
            Self::Info {} => "info".into(),
            Self::Measure {
                rpm: shot_rpm,
                landing_distance_m,
            } => {
                let distance = finite(*landing_distance_m, "landing_distance_m")?;
                if distance <= 0.0 || distance > 60.0 {
                    return Err(format!(
                        "landing_distance_m must be between 0 and 60, got {distance}"
                    ));
                }
                format!("measure {:.0} {:.4}", rpm(*shot_rpm)?, distance)
            }
            Self::Undo {} => "undo".into(),
            Self::Fit { height_m, kind } => {
                let height = finite(*height_m, "height_m")?;
                if height <= 0.0 || height > 5.0 {
                    return Err(format!("height_m must be between 0 and 5, got {height}"));
                }
                format!("fit {:.4} {}", height, kind.as_str())
            }
        })
    }

    /// True for anything that can move the machine or release a ball. Used for
    /// the evidence trail: an `info` poll is not an actuation and must not read
    /// like one.
    pub fn is_actuating(&self) -> bool {
        matches!(
            self,
            Self::Aim { .. }
                | Self::Wheels { .. }
                | Self::Reload {}
                | Self::Fire { .. }
                | Self::Center {}
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn parse(json: serde_json::Value) -> Result<ConsoleCommand, serde_json::Error> {
        serde_json::from_value(json)
    }

    fn render(json: serde_json::Value) -> Result<String, String> {
        parse(json).map_err(|error| error.to_string())?.render()
    }

    #[test]
    fn every_intent_renders_the_line_the_bridge_parses() {
        let cases = [
            (
                serde_json::json!({"command": "aim", "pitch_deg": 12.0,
                                   "yaw_deg": -7.0, "wheel_rpm": 800}),
                "aim 12 -7 800",
            ),
            (
                serde_json::json!({"command": "wheels", "wheel_rpm": 500}),
                "wheels 500",
            ),
            (serde_json::json!({"command": "reload"}), "reload"),
            (serde_json::json!({"command": "arm"}), "arm"),
            (serde_json::json!({"command": "disarm"}), "disarm"),
            (
                serde_json::json!({"command": "fire", "confirm": "FIRE"}),
                "fire",
            ),
            (serde_json::json!({"command": "stop"}), "stop"),
            (serde_json::json!({"command": "clear"}), "clear"),
            (serde_json::json!({"command": "center"}), "center"),
            (serde_json::json!({"command": "set_zero"}), "set_zero"),
            (
                serde_json::json!({"command": "limits", "pitch_min_deg": -25.0,
                                   "pitch_max_deg": 30.0}),
                "limits -25 30",
            ),
            (serde_json::json!({"command": "info"}), "info"),
            (
                serde_json::json!({"command": "measure", "rpm": 800,
                                   "landing_distance_m": 3.945}),
                "measure 800 3.9450",
            ),
            (serde_json::json!({"command": "undo"}), "undo"),
            (
                serde_json::json!({"command": "fit", "height_m": 0.52,
                                   "kind": "linear"}),
                "fit 0.5200 linear",
            ),
        ];
        for (payload, expected) in cases {
            assert_eq!(render(payload.clone()).unwrap(), expected, "{payload:?}");
        }
    }

    #[test]
    fn a_shot_needs_the_exact_confirmation_word() {
        assert!(render(serde_json::json!({"command": "fire", "confirm": "FIRE"})).is_ok());
        for wrong in ["", "fire", "Fire", "YES", "FIRE "] {
            assert!(
                render(serde_json::json!({"command": "fire", "confirm": wrong})).is_err(),
                "accepted confirm {wrong:?}"
            );
        }
        // A payload with no confirmation at all cannot be a shot.
        assert!(parse(serde_json::json!({"command": "fire"})).is_err());
    }

    #[test]
    fn out_of_range_geometry_and_rpm_are_refused_not_clamped() {
        // The firmware reboots beyond +/-30, so a UI asking for 45 is a bug that
        // must surface rather than quietly becoming a 30-degree shot.
        for bad in [
            serde_json::json!({"command": "aim", "pitch_deg": 45.0,
                               "yaw_deg": 0.0, "wheel_rpm": 800}),
            serde_json::json!({"command": "aim", "pitch_deg": 0.0,
                               "yaw_deg": -31.0, "wheel_rpm": 800}),
            serde_json::json!({"command": "aim", "pitch_deg": 0.0,
                               "yaw_deg": 0.0, "wheel_rpm": 1600}),
            serde_json::json!({"command": "aim", "pitch_deg": 0.0,
                               "yaw_deg": 0.0, "wheel_rpm": -1}),
            serde_json::json!({"command": "wheels", "wheel_rpm": 1201}),
            serde_json::json!({"command": "measure", "rpm": 800,
                               "landing_distance_m": 0.0}),
            serde_json::json!({"command": "measure", "rpm": 800,
                               "landing_distance_m": 61.0}),
            serde_json::json!({"command": "fit", "height_m": 0.0, "kind": "linear"}),
            serde_json::json!({"command": "fit", "height_m": 6.0, "kind": "linear"}),
        ] {
            assert!(render(bad.clone()).is_err(), "accepted {bad:?}");
        }
        // Exactly at the limits is legitimate operating range.
        assert!(
            render(serde_json::json!({"command": "aim", "pitch_deg": 30.0,
                                          "yaw_deg": -30.0, "wheel_rpm": 1200}))
            .is_ok()
        );
    }

    #[test]
    fn this_layer_guards_the_firmware_bound_and_leaves_the_envelope_to_the_bridge() {
        // REWRITTEN 2026-08-06, twice in one day, and the second version is the
        // one that is actually right.
        //
        // First attempt: refuse pitch < 0 here, after an operator drove the barrel
        // into the ball feeder. That looked like the fix and was the wrong frame —
        // the collision is at a fixed PHYSICAL position, while this angle is
        // measured from a zero adopted at boot or by `set_zero`. A constant in
        // this frame points somewhere else after every re-zero, and it also
        // deleted legitimate downward travel: how much room is left below zero
        // depends on where zero was put, which only the operator can see.
        //
        // So the division is: this layer guards +/-30, which is frame-independent
        // because it is about the ESP32 rebooting. The mechanical envelope is
        // per-session state in the bridge, declared with `limits` and clamped
        // there.
        for inside in [-30.0, -12.0, 0.0, 15.0, 30.0] {
            assert!(
                render(serde_json::json!({
                    "command": "aim", "pitch_deg": inside, "yaw_deg": 0.0, "wheel_rpm": 0
                }))
                .is_ok(),
                "{inside} is inside the firmware bound and must reach the bridge"
            );
        }
        for outside in [-30.5, 31.0, 45.0] {
            let refused = render(serde_json::json!({
                "command": "aim", "pitch_deg": outside, "yaw_deg": 0.0, "wheel_rpm": 0
            }))
            .unwrap_err();
            assert!(refused.contains("firmware bound"), "{outside}: {refused}");
        }

        // A declared envelope is bounded by the same firmware limit and must be
        // ordered — an inverted one would silently clamp every angle to a point.
        assert!(render(serde_json::json!({
            "command": "limits", "pitch_min_deg": -25.0, "pitch_max_deg": 30.0
        }))
        .is_ok());
        for bad in [
            serde_json::json!({"command": "limits", "pitch_min_deg": -31.0,
                               "pitch_max_deg": 30.0}),
            serde_json::json!({"command": "limits", "pitch_min_deg": 0.0,
                               "pitch_max_deg": 31.0}),
            serde_json::json!({"command": "limits", "pitch_min_deg": 10.0,
                               "pitch_max_deg": 10.0}),
            serde_json::json!({"command": "limits", "pitch_min_deg": 20.0,
                               "pitch_max_deg": 5.0}),
            serde_json::json!({"command": "limits", "pitch_min_deg": 10.0,
                               "pitch_max_deg": 20.0}),
            serde_json::json!({"command": "limits", "pitch_min_deg": -20.0,
                               "pitch_max_deg": -10.0}),
        ] {
            assert!(render(bad.clone()).is_err(), "accepted {bad:?}");
        }
    }

    #[test]
    fn non_finite_and_smuggled_fields_are_refused() {
        for raw in [
            r#"{"command":"aim","pitch_deg":NaN,"yaw_deg":0,"wheel_rpm":800}"#,
            r#"{"command":"aim","pitch_deg":Infinity,"yaw_deg":0,"wheel_rpm":800}"#,
            r#"{"command":"aim","pitch_deg":1e400,"yaw_deg":0,"wheel_rpm":800}"#,
            // No raw passthrough, under any key.
            r#"{"command":"stop","raw":"shoot"}"#,
            r#"{"command":"info","line":"set 0 0 900 900"}"#,
            r#"{"command":"aim","pitch_deg":0,"yaw_deg":0,"wheel_rpm":800,"extra":1}"#,
            // Unknown verbs cannot reach the bridge.
            r#"{"command":"shoot"}"#,
            r#"{"command":"setzero"}"#,
            r#"{"command":"jf500"}"#,
            // Duplicate keys must not resolve by last-wins.
            r#"{"command":"fire","confirm":"no","confirm":"FIRE"}"#,
            // Case-sensitive vocabulary.
            r#"{"command":"AIM","pitch_deg":0,"yaw_deg":0,"wheel_rpm":0}"#,
        ] {
            let parsed = serde_json::from_str::<ConsoleCommand>(raw);
            let refused = match parsed {
                Err(_) => true,
                Ok(command) => command.render().is_err(),
            };
            assert!(refused, "accepted adversarial console payload: {raw}");
        }
    }

    #[test]
    fn a_rendered_line_is_always_one_line_so_two_commands_cannot_be_smuggled() {
        // The bridge reads one intent per line. If any rendering could contain a
        // newline, a single approved intent could deliver a second, unapproved
        // one — so this holds for every variant, including the string-carrying
        // `fire` and the float formatting.
        let every = [
            serde_json::json!({"command": "aim", "pitch_deg": 30.0,
                               "yaw_deg": -30.0, "wheel_rpm": 950}),
            serde_json::json!({"command": "wheels", "wheel_rpm": 0}),
            serde_json::json!({"command": "reload"}),
            serde_json::json!({"command": "arm"}),
            serde_json::json!({"command": "disarm"}),
            serde_json::json!({"command": "fire", "confirm": "FIRE"}),
            serde_json::json!({"command": "stop"}),
            serde_json::json!({"command": "clear"}),
            serde_json::json!({"command": "center"}),
            serde_json::json!({"command": "set_zero"}),
            serde_json::json!({"command": "limits", "pitch_min_deg": -30.0,
                               "pitch_max_deg": 30.0}),
            serde_json::json!({"command": "info"}),
            serde_json::json!({"command": "measure", "rpm": 500,
                               "landing_distance_m": 2.4}),
            serde_json::json!({"command": "undo"}),
            serde_json::json!({"command": "fit", "height_m": 0.52,
                               "kind": "interp"}),
        ];
        for payload in every {
            let line = render(payload.clone()).unwrap();
            assert!(!line.contains('\n') && !line.contains('\r'), "{line:?}");
            assert_eq!(line.trim(), line, "{line:?}");
            assert!(!line.is_empty());
        }
    }

    #[test]
    fn only_machine_moving_intents_count_as_actuation() {
        let actuating = [
            serde_json::json!({"command": "aim", "pitch_deg": 0.0,
                               "yaw_deg": 0.0, "wheel_rpm": 800}),
            serde_json::json!({"command": "wheels", "wheel_rpm": 800}),
            serde_json::json!({"command": "reload"}),
            serde_json::json!({"command": "fire", "confirm": "FIRE"}),
            serde_json::json!({"command": "center"}),
        ];
        for payload in actuating {
            assert!(
                parse(payload.clone()).unwrap().is_actuating(),
                "{payload:?}"
            );
        }
        // Arming is a permission change, and stop/clear/info/bookkeeping move
        // nothing — recording them as actuation would make the evidence lie.
        for payload in [
            serde_json::json!({"command": "arm"}),
            serde_json::json!({"command": "disarm"}),
            serde_json::json!({"command": "stop"}),
            serde_json::json!({"command": "clear"}),
            serde_json::json!({"command": "set_zero"}),
            serde_json::json!({"command": "limits", "pitch_min_deg": 0.0,
                               "pitch_max_deg": 30.0}),
            serde_json::json!({"command": "info"}),
            serde_json::json!({"command": "measure", "rpm": 800,
                               "landing_distance_m": 3.9}),
            serde_json::json!({"command": "undo"}),
            serde_json::json!({"command": "fit", "height_m": 0.52,
                               "kind": "linear"}),
        ] {
            assert!(
                !parse(payload.clone()).unwrap().is_actuating(),
                "{payload:?}"
            );
        }
    }
}
